data-analytics/iot-nirvana/pipeline/src/test/java/com/google/cloud/demo/iot/nirvana/pipeline/FilterMessagesTest.java

```diff
@@ -0,0 +1,215 @@
+        CityTemperature.class);
+
+    cities = TemperaturePipeline.loadCitiesSideInput(pipeline);
+    goodTag = new TupleTag<KV<Message, CityTemperature>>() {};
+    errorTag = new TupleTag<String>() {};
+  }
+
+  /**
+   * Test {@link FilterMessages.processElement} with an invalid message, whose identifier does not
+   * correspond to the coordinates of an actual city.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void methodUnderTest_expectedResult() throws Exception {
+    // Create test data
+    List<String> inputMessages = Arrays.asList(invalidMessage);
+
+    // Create pipeline graph
+    PCollectionTuple filteredMessages = pipeline
+        .apply("Create", Create.of(inputMessages))
+        .apply(
+          "FilterMessages",
+          ParDo
+            .of(FilterMessages.newBuilder().setCities(cities).setErrorTag(errorTag).build())
+            .withSideInputs(cities)
+            .withOutputTags(goodTag, TupleTagList.of(errorTag)));
+    PCollection<KV<Message, CityTemperature>> validTemperatures = filteredMessages.get(goodTag);
+    PCollection<String> invalidTemperatures = filteredMessages.get(errorTag);
+
+    // Run pipeline and evaluate results
+    PAssert.that(validTemperatures).empty();
+    PAssert.that(invalidTemperatures).containsInAnyOrder(inputMessages);
+    pipeline.run();
+  }
+
+  /**
+   * Test {@link FilterMessages.processElement} with an valid message, whose identifier corresponds
+   * to the coordinates of an actual city.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testProcessElement_singleValidMessage() throws Exception {
+    // Create test data
+    List<String> inputMessages = Arrays.asList(validMessageTiranaStr);
+    List<KV<Message, CityTemperature>> expectedResult = Arrays.asList(
+        KV.of(validMessageTirana, cityTirana));
+
+    // Create pipeline graph
+    PCollectionTuple filteredMessages = pipeline
+        .apply("Create", Create.of(inputMessages))
+        .apply(
+          "FilterMessages",
+          ParDo
+            .of(FilterMessages.newBuilder().setCities(cities).setErrorTag(errorTag).build())
+            .withSideInputs(cities)
+            .withOutputTags(goodTag, TupleTagList.of(errorTag)));
+    PCollection<KV<Message, CityTemperature>> validTemperatures = filteredMessages.get(goodTag);
+    PCollection<String> invalidTemperatures = filteredMessages.get(errorTag);
+
+    // Run pipeline and evaluate results
+    PAssert.that(validTemperatures).containsInAnyOrder(expectedResult);
+    PAssert.that(invalidTemperatures).empty();
+    pipeline.run();
+  }
+
+  /**
+   * Test {@link FilterMessages.processElement} with multiple valid messages, whose respective
+   * identifiers corresponds to coordinates of an actual cities.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testProcessElement_multipleValidMessages() throws Exception {
```

Comment:
https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names

I think its better to either use either processElement_multipleValidMessages 
OR
testMultipleValidMessages

This comment applies to the other test cases as well.

Reply:
Changed all test method names according to style guide in commit 7077a410843d5d3bf4c5220b7e3c4582f9576d6a